mlflow/models/utils.py

```diff
@@ -270,89 +282,118 @@ def __init__(self, input_example: ModelInputExample):
+                        "- scipy.sparse.csc_matrix\n"
+                        "- dict\n"
+                        "- list\n"
+                        "- scalars\n"
+                        f"but got '{type(input_example)}'",
+                    )
+                self.data = _convert_dataframe_to_split_dict(self._inference_data)
+                self.serving_input = {DF_SPLIT: self.data}
+                orient = "split" if "columns" in self.data else "values"
                 self.info.update(
                     {
                         "type": "dataframe",
-                        "pandas_orient": "split",
+                        "pandas_orient": orient,
                     }
                 )
-        else:
-            self._inference_data = _coerce_to_pandas_df(input_example)
-            if self._inference_data is None:
-                raise TypeError(
-                    "Expected one of the following types:\n"
-                    "- pandas.DataFrame\n"
-                    "- numpy.ndarray\n"
-                    "- dictionary of (name -> numpy.ndarray)\n"
-                    "- scipy.sparse.csr_matrix\n"
-                    "- scipy.sparse.csc_matrix\n"
-                    "- dict\n"
-                    "- list\n"
-                    "- scalars\n"
-                    f"but got '{type(input_example)}'",
-                )
-            self.data = _handle_dataframe_input(self._inference_data)
-            orient = "split" if "columns" in self.data else "values"
-            self.info.update(
-                {
-                    "type": "dataframe",
-                    "pandas_orient": orient,
-                }
-            )
 
+    # TODO: isolate to make sure `artifact_path` saves an example that can be
+    # directly passed to model.predict. `serving_input_path` saves an example that
+    # can be directly passed to model serving inference.
     def save(self, parent_dir_path: str):
-        """Save the example as json at ``parent_dir_path``/`self.info['artifact_path']`."""
+        """
+        Save the example as json at ``parent_dir_path``/`self.info['artifact_path']`.
+        Save serving input as json at ``parent_dir_path``/`self.info['serving_input_path']`.
+        """
         if self._inference_params is not None:
             data = {EXAMPLE_DATA_KEY: self.data, EXAMPLE_PARAMS_KEY: self._inference_params}
+            serving_input = {**self.serving_input, SERVING_PARAMS_KEY: self._inference_params}
         else:
             data = self.data
+            serving_input = self.serving_input
         with open(os.path.join(parent_dir_path, self.info[INPUT_EXAMPLE_PATH]), "w") as f:
             json.dump(data, f, cls=NumpyEncoder)
+        if serving_input:
+            self.info[SERVING_INPUT_PATH] = SERVING_INPUT_FILENAME
+            with open(os.path.join(parent_dir_path, self.info[SERVING_INPUT_PATH]), "w") as f:
+                json.dump(serving_input, f, cls=NumpyEncoder)
```

Comment:
Can we do a validation that numpy missing data handlers generate JSON-compatible-for-model-serving example data to be sure that we cover as many compatibility concerns for a user copying this generated example and submitting it to a deployed endpoint? (just to be safe)